File: pattern-diagnose/assets/examples/go/factory-method/after/export_service.go
// SPDX-License-Identifier: MIT
// After: extracted Factory Method — creation logic deferred to factory function

package factorymethod

import (
	"encoding/json"
	"fmt"
)

// Document is the data to export.
type Document struct {
	Title   string
	Content string
}

// Exporter is the product interface. In Go, satisfied implicitly.
type Exporter interface {
	Export(doc Document) (string, error)
}

type JsonExporter struct{}

func (e *JsonExporter) Export(doc Document) (string, error) {
	b, err := json.Marshal(map[string]string{"title": doc.Title, "body": doc.Content})
	return string(b), err
}

type CsvExporter struct{}

func (e *CsvExporter) Export(doc Document) (string, error) {
	return fmt.Sprintf(`"title","content"%s"%s","%s"`, "\n", doc.Title, doc.Content), nil
}

type XmlExporter struct{}

func (e *XmlExporter) Export(doc Document) (string, error) {
	return fmt.Sprintf("<doc><title>%s</title><body>%s</body></doc>", doc.Title, doc.Content), nil
}

// Factory function — sufficient when creation logic is stateless.
// Go's implicit interface satisfaction means no class hierarchy needed.
func NewExporter(format string) (Exporter, error) {
	switch format {
	case "json":
		return &JsonExporter{}, nil
	case "csv":
		return &CsvExporter{}, nil
	case "xml":
		return &XmlExporter{}, nil
	default:
		return nil, fmt.Errorf("unknown format: %s", format)
	}
}

// Client code no longer constructs exporters directly.
type ExportService struct{}

func (s *ExportService) Export(doc Document, format string) (string, error) {
	exporter, err := NewExporter(format)
	if err != nil {
		return "", err
	}
	return exporter.Export(doc)
}

File: pattern-diagnose/assets/examples/typescript/factory-method/after/export-service.ts
// SPDX-License-Identifier: MIT
// After: extracted Factory Method — creation logic deferred to creator subtypes

interface Document {
  title: string;
  content: string;
}

interface Exporter {
  export(doc: Document): string;
}

class JsonExporter implements Exporter {
  export(doc: Document): string {
    return JSON.stringify({ title: doc.title, body: doc.content });
  }
}

class CsvExporter implements Exporter {
  export(doc: Document): string {
    return `"title","content"\n"${doc.title}","${doc.content}"`;
  }
}

class XmlExporter implements Exporter {
  export(doc: Document): string {
    return `<doc><title>${doc.title}</title><body>${doc.content}</body></doc>`;
  }
}

// Factory function — sufficient when creation logic is stateless
// (No need for a Creator class hierarchy in TypeScript)
function createExporter(format: string): Exporter {
  switch (format) {
    case "json":
      return new JsonExporter();
    case "csv":
      return new CsvExporter();
    case "xml":
      return new XmlExporter();
    default:
      throw new Error(`Unknown format: ${format}`);
  }
}

// Client code no longer constructs exporters directly
class ExportService {
  export(doc: Document, format: string): string {
    const exporter = createExporter(format);
    return exporter.export(doc);
  }
}
